lms/security.py

```diff
@@ -212,13 +213,24 @@ def _get_lti_jwt(request):
     if not id_token:
         return {}
 
-    try:
-        jwt_params = jwt.decode(id_token, options={"verify_signature": False})
-    except InvalidTokenError as err:
-        raise marshmallow.ValidationError("Invalid id_token", "authorization") from err
+    jwt_service = request.find_service(JWTService)
+    registration_service = request.find_service(LTIRegistrationService)
 
-    warnings.warn("Using not verified JWT token")
-    return jwt_params
+    # Get the target registration of the JWT.
+    # We'll use the public key URL in the registration to verify the token
+    unverified_payload = jwt_service.decode_unverified(id_token)
+    registration = registration_service.get(
+        unverified_payload.get("iss"), unverified_payload.get("aud")
+    )
+    if not registration:
+        return {}
+
+    try:
+        return request.find_service(JWTService).decode_with_jwk_url(
+            id_token, jwk_url=registration.key_set_url, audience=registration.client_id
+        )
+    except InvalidJWTError:
+        return {}
```

Comment:
Moved most of this to the service. Left the logging here so the service only has to deal with rising the exceptions.

Reply:
I feel like this is kind of doing a smaller version of the situation before? The method is only called once (here) so why not have it behave the correct way instead of catching it etc. It's ok to log in our services.

At the moment we have a raise which we test, another test testing we catch it. There's a similar story (but smaller as it doesn't involve the service) if `id_token` is null.

I feel like this method could be:

```python
def _get_lti_jwt(request):
    return request.find_service(JWTService).decode_lti_token(
        id_token=request.params.get("id_token")
    )
```

If you look at the `_get_user` above it's exactly how that's done too.

I should have thought of this before, but looking at this I'm not sure why the attachment of the request method shouldn't happen in the service too for maximal localism? At the moment if you want to find out where this comes from in `lti_launch` for example. You grep for it, and find it in security, but none of the contents of the method are here, so you immediately get redirected off to `JWTService`.

The same is true for `_get_user`. Putting these side by side would mean they are right next to the code they call and also that as much of the `JWT` stuff is in the same place as possible.